ath79: add support for jjPlus JWAP230

ovalenti opened the pull request:
ath79: add support for jjPlus JWAP230


The jjPlus JWAP230 is an access point board built around the QCA9558,
with built-in 2.4GHz 3x3 N WiFi (28dBm). It can be expanded with 2
mini-PCIe boards, and has an USB2 root port.

Specifications:
- SOC: Qualcomm Atheros QCA9558
- CPU: 720MHz
- H/W switch: QCA8327 rev 2
- Flash: 16 MiB SPI NOR (en25qh128)
- RAM: 128 MiB DDR2
- WLAN: AR9550 built-in SoC bgn 3T3R (ath9k)
- PCI: 2x mini-PCIe (optional 5V)
- USB2:
  - 1x Type A root port
  - 1x combined mini-PCIe
- Ethernet:
  - 2x 10/100/1000 (1x PoE 802.3af (36-57 V))

Notes:
 The device used to be supported in the ar71xx target.

Flash instructions:
- install from u-boot with tftp (requires serial access)
  > setenv ipaddr a.b.c.d
  > setenv serverip e.f.g.h
  > tftp 0x80060000 \
      openwrt-ath79-generic-jjplus_jwap230-squashfs-sysupgrade.bin
  > erase 0x9f050000 +${filesize}
  > cp.b $fileaddr 0x9f050000 $filesize
  > setenv bootcmd bootm 0x9f050000
  > saveenv

Signed-off-by: Olivier Valentin <valentio@free.fr>

adschm commented on target/linux/ath79/dts/qca9558_jjplus_jwap230.dts:
```diff
@@ -0,0 +1,165 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+
+/dts-v1/;
```
dts-v1 can be dropped and qca955x.dtsi should be moved before the other includes.

ovalenti replied:
The file was updated accordingly.

adschm commented on target/linux/ath79/dts/qca9558_jjplus_jwap230.dts:
```diff
@@ -0,0 +1,165 @@
+
+	chosen {
+		bootargs = "console=ttyS0,115200n8";
+	};
```
This block can be dropped, it's already in the dtsi.

ovalenti replied:
Good point, sorry for that one

adschm commented on target/linux/ath79/dts/qca9558_jjplus_jwap230.dts:
```diff
@@ -0,0 +1,165 @@
+
+&gpio {
+	status = "okay";
+};
```
uart and gpio are both already enabled and can be dropped here.

ovalenti replied:
Done. I also removed the entire button section since I learned that the reset button is hardware.

adschm commented on target/linux/ath79/dts/qca9558_jjplus_jwap230.dts:
```diff
@@ -0,0 +1,165 @@
+
+&spi {
+	status = "okay";
+	num-cs = <1>;
```
num-cs should be dropped
https://github.com/openwrt/openwrt/commit/72bd92bea0aacf9a008782bea9e015c3ac0a2888

ovalenti replied:
Done
Thanks for the pointer.

adschm commented on target/linux/ath79/dts/qca9558_jjplus_jwap230.dts:
```diff
@@ -0,0 +1,165 @@
+				read-only;
+			};
+
+			partition@20000 {
```
this does not match the offset in the reg property. same for the following node.

ovalenti replied:
I updated the node names to reflect the offsets

adschm commented on target/linux/ath79/dts/qca9558_jjplus_jwap230.dts:
```diff
@@ -0,0 +1,165 @@
+			};
+
+			art: partition@ff0000 {
+				label = "ART";
```
This is typically lower-case in ath79.

adschm commented on target/linux/ath79/dts/qca9558_jjplus_jwap230.dts:
```diff
@@ -0,0 +1,165 @@
+&eth0 {
+	status = "okay";
+
+	mtd-mac-address = <&art 0x00>;
```
Please use nvmem-cells instead.

adschm commented on target/linux/ath79/dts/qca9558_jjplus_jwap230.dts:
```diff
@@ -0,0 +1,165 @@
+
+&mdio1 {
+	status = "disabled";
+};
```
This is disabled by default?

adschm commented on target/linux/ath79/dts/qca9558_jjplus_jwap230.dts:
```diff
@@ -0,0 +1,165 @@
+	status = "okay";
+
+	mtd-mac-address = <&art 0x00>;
+	mtd-mac-address-increment = <1>;
```
nvmem-cells

adschm commented on target/linux/ath79/dts/qca9558_jjplus_jwap230.dts:
```diff
@@ -0,0 +1,165 @@
+	fixed-link {
+		speed = <1000>;
+		full-duplex;
+	};
```
Does having fixed-link twice make sense if you use the switch?

ovalenti replied:
Since there are 2 distinct MII buses in use in this machine, I believe that this is required to configure both of them.

eth0 is RGMII connected to PAD0 of the switch
eth1 is SGMII connected to PAD6 of the switch

adschm commented on target/linux/ath79/dts/qca9558_jjplus_jwap230.dts:
```diff
@@ -0,0 +1,165 @@
+	status = "okay";
+	mtd-cal-data = <&art 0x1000>;
+	mtd-mac-address = <&art 0x00>;
+	mtd-mac-address-increment = <(-2)>;
```
nvmem-cells ...

Are there valid addresses in the other common locations? 0x6, 0x1002 etc.?

Please add a MAC address overview to your commit message:
https://openwrt.org/docs/guide-developer/device-support-policies#mac_addresses

ovalenti replied:
I could find the MAC addresses matching the device label in the dump of the ART partition; right where you pointed.

adschm commented on target/linux/ath79/dts/qca9558_jjplus_jwap230.dts:
```diff
@@ -0,0 +1,165 @@
+&usb1 {
+	status = "okay";
+};
+
```
Remove empty line at EOF.

ovalenti replied:
I think it is ok now.

adschm commented on target/linux/ath79/image/generic.mk:
```diff
@@ -1398,6 +1398,16 @@ define Device/jjplus_ja76pf2
+  SOC := qca9558
+  DEVICE_VENDOR := jjPlus
+  DEVICE_MODEL := JWAP230
+  KERNEL := kernel-bin | append-dtb | lzma | uImage lzma
```
This is the same as in Device/Default and thus can be dropped ...

ovalenti replied:
I switched to the default sysupgrade rule. In the end, the image definition is rather minimalistic now.

adschm commented on target/linux/ath79/image/generic.mk:
```diff
@@ -1398,6 +1398,16 @@ define Device/jjplus_ja76pf2
+  DEVICE_MODEL := JWAP230
+  KERNEL := kernel-bin | append-dtb | lzma | uImage lzma
+  IMAGES += factory.bin
+  IMAGE/factory.bin := append-kernel | pad-to $$(BLOCKSIZE) | append-rootfs | pad-rootfs | check-size 16000k
```
Please wrap after 74 chars. max.

Please specify IMAGE_SIZE variable and use check-size without argument then.

ovalenti replied:
Let me know if I got it right, please ?

adschm reviewed (changes requested):
No LEDs?

adschm commented:
> - Ethernet: PoE 802.3af/at
>   - JWAP230: 2x 10/100/1000
>   - JWAP232: 3x 10/100/1000

So, is this actually adding support for two different devices? If yes, please add two different devices, each with the correct ethernet configuration.

ovalenti commented:
Thanks for the review ! As far as know, the only user LEDs available on jwap230 are controlled by the Switch chip, and the SoC has no possibility to use them.

Commit:
ath79: add support for jjPlus JWAP230

The jjPlus JWAP230 is an access point board built around the QCA9558,
with built-in 2.4GHz 3x3 N WiFi (28dBm). It can be expanded with 2
mini-PCIe boards, and has an USB2 root port.

Specifications:
- SOC: Qualcomm Atheros QCA9558
- CPU: 720MHz
- H/W switch: QCA8327 rev 2
- Flash: 16 MiB SPI NOR (en25qh128)
- RAM: 128 MiB DDR2
- WLAN: AR9550 built-in SoC bgn 3T3R (ath9k)
- PCI: 2x mini-PCIe (optional 5V)
- USB2:
  - 1x Type A root port
  - 1x combined mini-PCIe
- Ethernet:
  - 2x 10/100/1000 (1x PoE 802.3af (36-57 V))

Notes:
 The device used to be supported in the ar71xx target.

MAC address assignment:
    use   source
    LAN   art 0x0
    WAN   art 0x6
    WLAN  art 0x1002 (as part of the calibration data)

Flash instructions:
- install from u-boot with tftp (requires serial access)
  > setenv ipaddr a.b.c.d
  > setenv serverip e.f.g.h
  > tftp 0x80060000 \
      openwrt-ath79-generic-jjplus_jwap230-squashfs-sysupgrade.bin
  > erase 0x9f050000 +${filesize}
  > cp.b $fileaddr 0x9f050000 $filesize
  > setenv bootcmd bootm 0x9f050000
  > saveenv

Signed-off-by: Olivier Valentin <valentio@free.fr>

ovalenti commented:
You are right, @adschm, the JWAP230 and JWAP232 are two distinct models. They however differ only by the number of physical ethernet ports available.
Since these devices are rather exotic, I thought I would limit my additions to 20_network by making the both models share the same configuration.

For now, I removed all references to JWAP232, so that when I get this one right, I can make a new PR.
Unless you prefer having both in a row ?

chunkeey commented on target/linux/ath79/dts/qca9558_jjplus_jwap230.dts:
```diff
@@ -0,0 +1,159 @@
+
+&wmac {
+	status = "okay";
+	mtd-cal-data = <&art 0x1000>;
```
you can use the "calibration" nvmem-cell. (I.e: TP-Link Archer C7-v1 / qca9558_tplink_archer-c7-v1.dts)

ovalenti replied:
I updated this section to make use of the nvmem framework.

chunkeey replied:
Oh, you kept the mtd-cal-data around. You don't need to. You can delete it.

Also, I just realized: You can safely remove the mac-address cells (including the macaddr_art_wlan) for the wmac WLAN too! Reason is that the Mac-Address is part ot the calibration data. The driver will automatically set it correctly.

ovalenti replied:
Ow, I am sorry...
The mac address hint sounds logical; I will update this. Thanks

chunkeey commented on target/linux/ath79/dts/qca9558_jjplus_jwap230.dts:
```diff
@@ -0,0 +1,159 @@
+	nvmem-cell-names = "mac-address";
+};
+
+&art {
```
since this is all the same file, you might as well move the content of this into the "art" partition node (line 51 - 55) above.

ovalenti replied:
I was tempted to do so at first, but thought it would prevent overloading of the partitioning section.
If this is the usual best practice, I will be happy to move the nv info.
Thanks a lot for the review !

ovalenti commented:
@adschm , @chunkeey , I just updated the content of the PR based on your comments. I believe that they are all in order now.

I would be very grateful if you could check the new version of the files and tell me if there is anything I missed.

chunkeey commented:
Had to do some more things. The vendor also talks about a Reset button, do you know where it is?
Anyway, merged it for now.

ovalenti commented:
Unfortunately, the reset button is not a GPIO. It seems to be connected directly to the soc reset logic.

This device is more an industrial product than an end user thing. We use it in custom enclosures to build big batman networks.

Thanks a lot for your support and help getting this PR right!
